File: typescript/packages/providers/src/mock.ts
import { randomUUID } from 'node:crypto';
import type { ModelChunk } from '@agent-framework/core';
import type { ModelProvider, StreamRequest } from '@agent-framework/core';

export interface MockProviderOptions {
  responses?: Array<{ text?: string; toolCalls?: Array<{ name: string; arguments: Record<string, unknown> }> }>;
}

export class MockProvider implements ModelProvider {
  readonly id = 'mock';
  private turn = 0;

  constructor(private readonly options: MockProviderOptions = {}) {}

  async *stream(_request: StreamRequest): AsyncGenerator<ModelChunk, void, unknown> {
    const scripted = this.options.responses?.[this.turn];
    this.turn += 1;

    if (!scripted) {
      yield { type: 'text_delta', textDelta: 'Done.' };
      yield { type: 'usage', usage: { input: 1, output: 1, total: 2 } };
      yield { type: 'done' };
      return;
    }

    if (scripted.text) {
      yield { type: 'text_delta', textDelta: scripted.text };
    }

    for (const call of scripted.toolCalls ?? []) {
      yield {
        type: 'tool_call',
        toolCall: {
          id: randomUUID(),
          name: call.name,
          arguments: call.arguments,
        },
      };
    }

    yield { type: 'usage', usage: { input: 10, output: 5, total: 15 } };
    yield { type: 'done' };
  }
}

File: python/src/agent_framework/providers/mock.py
from __future__ import annotations

from typing import AsyncIterator
from uuid import uuid4

from agent_framework.core.domain import Message, ModelChunk, ToolCallRequest, ToolDefinition


class MockProvider:
    id = "mock"

    def __init__(self, responses: list[dict] | None = None) -> None:
        self._responses = responses or []
        self._turn = 0

    async def stream(self, messages: list[Message], tools: list[ToolDefinition]) -> AsyncIterator[ModelChunk]:
        del messages, tools
        scripted = self._responses[self._turn] if self._turn < len(self._responses) else None
        self._turn += 1

        if scripted is None:
            yield ModelChunk(type="text_delta", text_delta="Done.")
            yield ModelChunk(type="usage", usage={"input": 1, "output": 1, "total": 2})
            yield ModelChunk(type="done")
            return

        if scripted.get("text"):
            yield ModelChunk(type="text_delta", text_delta=scripted["text"])

        for call in scripted.get("tool_calls", []):
            yield ModelChunk(
                type="tool_call",
                tool_call=ToolCallRequest(
                    id=str(uuid4()),
                    name=call["name"],
                    arguments=call["arguments"],
                ),
            )

        yield ModelChunk(type="usage", usage={"input": 10, "output": 5, "total": 15})
        yield ModelChunk(type="done")
